google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1alpha2/JsonStreamWriter.java

```diff
@@ -0,0 +1,404 @@
+      @Nullable BatchingSettings batchingSettings,
+      @Nullable RetrySettings retrySettings,
+      @Nullable ExecutorProvider executorProvider,
+      @Nullable String endpoint) {
+    if (channelProvider != null) {
+      builder.setChannelProvider(channelProvider);
+    }
+    if (credentialsProvider != null) {
+      builder.setCredentialsProvider(credentialsProvider);
+    }
+    if (batchingSettings != null) {
+      builder.setBatchingSettings(batchingSettings);
+    }
+    if (retrySettings != null) {
+      builder.setRetrySettings(retrySettings);
+    }
+    if (executorProvider != null) {
+      builder.setExecutorProvider(executorProvider);
+    }
+    if (endpoint != null) {
+      builder.setEndpoint(endpoint);
+    }
+    JsonStreamWriterOnSchemaUpdateRunnable jsonStreamWriterOnSchemaUpdateRunnable =
+        new JsonStreamWriterOnSchemaUpdateRunnable();
+    jsonStreamWriterOnSchemaUpdateRunnable.setJsonStreamWriter(this);
+    builder.setOnSchemaUpdateRunnable(jsonStreamWriterOnSchemaUpdateRunnable);
+  }
+
+  /**
+   * Setter for table schema. Used for schema updates.
+   *
+   * @param tableSchema
+   */
+  void setTableSchema(Table.TableSchema tableSchema) {
+    this.tableSchema = tableSchema;
+  }
+
+  /**
+   * newBuilder that constructs a JsonStreamWriter builder with BigQuery client being initialized by
+   * StreamWriter by default.
+   *
+   * @param streamName name of the stream that must follow
+   *     "projects/[^/]+/datasets/[^/]+/tables/[^/]+/streams/[^/]+"
+   * @param tableSchema The schema of the table when the stream was created, which is passed back
+   *     through {@code WriteStream}
+   * @return Builder
+   */
+  public static Builder newBuilder(String streamName, Table.TableSchema tableSchema) {
+    Preconditions.checkNotNull(streamName, "StreamName is null.");
+    Preconditions.checkNotNull(tableSchema, "TableSchema is null.");
+    return new Builder(streamName, tableSchema, null);
+  }
+
+  /**
+   * newBuilder that constructs a JsonStreamWriter builder.
+   *
+   * @param streamName name of the stream that must follow
+   *     "projects/[^/]+/datasets/[^/]+/tables/[^/]+/streams/[^/]+"
+   * @param tableSchema The schema of the table when the stream was created, which is passed back
+   *     through {@code WriteStream}
+   * @param client
+   * @return Builder
+   */
+  public static Builder newBuilder(
+      String streamName, Table.TableSchema tableSchema, BigQueryWriteClient client) {
+    Preconditions.checkNotNull(streamName, "StreamName is null.");
+    Preconditions.checkNotNull(tableSchema, "TableSchema is null.");
+    Preconditions.checkNotNull(client, "BigQuery client is null.");
+    return new Builder(streamName, tableSchema, client);
+  }
+
+  /** Closes the underlying StreamWriter. */
+  @Override
+  public void close() {
+    this.streamWriter.close();
+  }
+
+  public class JsonStreamWriterOnSchemaUpdateRunnable extends OnSchemaUpdateRunnable {
```

Comment:
This can be defined as inner class to JsonStreamWriter. It is implementation detail.

Reply:
Got it.